contracts/external/cw-vesting/src/contract.rs

```diff
@@ -0,0 +1,549 @@
+            }
+        }
+    };
+
+    // Update vesting payment with amount of staked rewards
+    VESTING_PAYMENT.update(deps.storage, |mut vp| -> Result<_, ContractError> {
+        // Check stake amount is not greater than vesting payment amount
+        if amount > vp.amount {
+            return Err(ContractError::NotEnoughFunds {});
+        }
+
+        // Check that amount has not already been staked
+        vp.amount.checked_sub(vp.staked_amount + amount)?;
+
+        // Update amounts
+        vp.staked_amount += vp.staked_amount.checked_add(amount)?;
+        Ok(vp)
+    })?;
+
+    // Save a record of staking this vesting payment amount to a validator
+    match STAKED_VESTING_BY_VALIDATOR.may_load(deps.storage, &validator)? {
+        // If already staked to this validator, increase staked amount
+        Some(staked_amount) => STAKED_VESTING_BY_VALIDATOR.save(
+            deps.storage,
+            &validator,
+            &staked_amount.checked_add(amount)?,
+        ),
+        // If not currently staked to this validator save a new record with staked amount
+        None => STAKED_VESTING_BY_VALIDATOR.save(deps.storage, &validator, &amount),
+    }?;
+
+    // Create message to delegate the underlying tokens
+    let msg = StakingMsg::Delegate {
+        validator: validator.clone(),
+        amount: Coin { denom, amount },
+    };
+
+    Ok(Response::new()
+        .add_attribute("method", "delegate")
+        .add_attribute("amount", amount.to_string())
+        .add_attribute("validator", validator)
+        .add_message(msg))
+}
+
+pub fn execute_redelegate(
+    _env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    src_validator: String,
+    dst_validator: String,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    // Non-payable as this method redelegates funds in the contract
+    nonpayable(&info)?;
+
+    let vesting_payment = VESTING_PAYMENT.load(deps.storage)?;
+
+    // Check status is active
+    if vesting_payment.status != VestingPaymentStatus::Active {
+        return Err(ContractError::NotActive);
+    }
+
+    // Check sender is the recipient of the vesting payment
+    if info.sender != vesting_payment.recipient {
+        return Err(ContractError::Unauthorized);
+    }
+
+    // Check vesting payment denom matches local denom
+    let denom = match vesting_payment.denom {
+        CheckedDenom::Cw20(_) => {
+            return Err(ContractError::NotStakeable);
+        }
+        CheckedDenom::Native(denom) => {
+            // Get local denom, ensure it matches the vesting denom
+            let local_denom = deps.querier.query_bonded_denom()?;
+            if local_denom != denom {
+                return Err(ContractError::NotStakeable);
+            } else {
+                denom
+            }
+        }
+    };
+
+    // Load source validator, subtract redelegated amount
+    let amount_staked_src = STAKED_VESTING_BY_VALIDATOR.load(deps.storage, &src_validator)?;
```

Comment:
we could map this error to a more informational error that src validator is not delegated to